File: src/scraper/controller/protocol.py
"""Message shapes, per SPEC-V3's "Messages" and "Protocol invariants"
sections. Every message carries `type` and `protocolVersion`; nothing here
adds a message type that isn't in SPEC-V3's table.

Extension-supplied messages are untrusted input (SPEC-V3, Input hardening)
and are validated field-by-field on decode. Controller-supplied messages are
built by typed constructors here so a caller can't accidentally omit a
required field or add one that isn't in the spec.
"""

import json
from typing import Any

PROTOCOL_VERSION = 1

ERROR_CODES = frozenset(
    {
        "PAGE_READY_TIMEOUT",
        "CONSENT_WALL",
        "VIDEO_UNAVAILABLE",
        "AGE_RESTRICTED",
        "LOGIN_REQUIRED",
        "SHORTS_EXCLUDED",
        "UNEXPECTED_NAVIGATION",
        "SCHEMA_UNRECOGNISED",
        "CHAIN_STALLED",
        "CIRCUIT_BREAKER",
        "PAYLOAD_TOO_LARGE",
        "CONTROLLER_UNAVAILABLE",
    }
)

COMPLETION_REASONS = frozenset({"max_recommendations_reached", "chain_exhausted"})
PAYLOAD_SOURCES = frozenset({"initial", "network"})

_EXTENSION_MESSAGE_FIELDS = {
    "hello": frozenset({"restart"}),
    "payload": frozenset(
        {"runId", "videoId", "navEpoch", "batchId", "source", "endpoint", "capturedAt", "body"}
    ),
    "video_done": frozenset({"runId", "videoId", "navEpoch", "reason"}),
    "video_failed": frozenset({"runId", "videoId", "navEpoch", "errorCode", "message"}),
    "ping": frozenset({"runId"}),
}


class ProtocolError(Exception):
    """Raised for any malformed or invalid message from the extension. The
    caller should treat the connection as unrecoverable — untrusted input
    that fails validation is not something to guess a recovery for."""


class ProtocolVersionMismatch(ProtocolError):
    pass


def decode_extension_message(raw: bytes) -> dict[str, Any]:
    """Parse and validate one frame's bytes as an extension-to-controller
    message. Raises ProtocolError on anything that doesn't match SPEC-V3's
    table exactly."""
    try:
        msg = json.loads(raw.decode("utf-8"))
    except (json.JSONDecodeError, UnicodeDecodeError) as exc:
        raise ProtocolError(f"invalid JSON: {exc}") from exc

    if not isinstance(msg, dict):
        raise ProtocolError(f"message must be a JSON object, got {type(msg).__name__}")

    msg_type = msg.get("type")
    if msg_type not in _EXTENSION_MESSAGE_FIELDS:
        raise ProtocolError(f"unknown or missing message type: {msg_type!r}")

    if "protocolVersion" not in msg:
        raise ProtocolError("message is missing protocolVersion")

    required = _EXTENSION_MESSAGE_FIELDS[msg_type]
    missing = required - msg.keys()
    if missing:
        raise ProtocolError(f"{msg_type} message missing fields: {sorted(missing)}")

    if msg_type == "payload":
        if msg["source"] not in PAYLOAD_SOURCES:
            raise ProtocolError(f"payload has unknown source: {msg['source']!r}")
    elif msg_type == "video_done":
        if msg["reason"] not in COMPLETION_REASONS:
            raise ProtocolError(f"video_done has unknown reason: {msg['reason']!r}")
    elif msg_type == "video_failed":
        if msg["errorCode"] not in ERROR_CODES:
            raise ProtocolError(f"video_failed has unknown errorCode: {msg['errorCode']!r}")

    return msg


def encode_message(msg: dict[str, Any]) -> bytes:
    return json.dumps(msg).encode("utf-8")


def check_protocol_version(msg: dict[str, Any]) -> None:
    if msg["protocolVersion"] != PROTOCOL_VERSION:
        raise ProtocolVersionMismatch(
            f"protocol version mismatch: extension sent {msg['protocolVersion']!r}, "
            f"controller expects {PROTOCOL_VERSION!r}"
        )


# -- Controller -> extension message constructors ---------------------------


def build_hello_ack(run_id: str, video_id: str, video_url: str, config: dict) -> dict:
    return {
        "type": "hello_ack",
        "protocolVersion": PROTOCOL_VERSION,
        "runId": run_id,
        "videoId": video_id,
        "videoUrl": video_url,
        "config": config,
    }


def build_payload_ack(batch_id: str) -> dict:
    return {"type": "payload_ack", "protocolVersion": PROTOCOL_VERSION, "batchId": batch_id}


def build_payload_nack(batch_id: str, retryable: bool) -> dict:
    return {
        "type": "payload_nack",
        "protocolVersion": PROTOCOL_VERSION,
        "batchId": batch_id,
        "retryable": retryable,
    }


def build_video_done_ack(video_id: str) -> dict:
    return {"type": "video_done_ack", "protocolVersion": PROTOCOL_VERSION, "videoId": video_id}


def build_video_failed_ack(video_id: str) -> dict:
    return {"type": "video_failed_ack", "protocolVersion": PROTOCOL_VERSION, "videoId": video_id}


def build_pong() -> dict:
    return {"type": "pong", "protocolVersion": PROTOCOL_VERSION}


def build_stop() -> dict:
    return {"type": "stop", "protocolVersion": PROTOCOL_VERSION}

File: extension/src/protocol.ts
/**
 * Message shapes, mirroring src/scraper/controller/protocol.py exactly.
 * Nothing here adds a message type or field that isn't in SPEC-V3's
 * "Messages" table.
 */

export const PROTOCOL_VERSION = 1;

export type ErrorCode =
  | "PAGE_READY_TIMEOUT"
  | "CONSENT_WALL"
  | "VIDEO_UNAVAILABLE"
  | "AGE_RESTRICTED"
  | "LOGIN_REQUIRED"
  | "SHORTS_EXCLUDED"
  | "UNEXPECTED_NAVIGATION"
  | "SCHEMA_UNRECOGNISED"
  | "CHAIN_STALLED"
  | "CIRCUIT_BREAKER"
  | "PAYLOAD_TOO_LARGE"
  | "CONTROLLER_UNAVAILABLE";

export type CompletionReason = "max_recommendations_reached" | "chain_exhausted";
export type PayloadSource = "initial" | "network";

// -- Extension -> controller --------------------------------------------

export interface HelloMessage {
  type: "hello";
  protocolVersion: number;
  restart: boolean;
}

export interface PayloadMessage {
  type: "payload";
  protocolVersion: number;
  runId: string;
  videoId: string;
  navEpoch: number;
  batchId: string;
  source: PayloadSource;
  endpoint: string;
  capturedAt: string;
  body: unknown;
}

export interface VideoDoneMessage {
  type: "video_done";
  protocolVersion: number;
  runId: string;
  videoId: string;
  navEpoch: number;
  reason: CompletionReason;
}

export interface VideoFailedMessage {
  type: "video_failed";
  protocolVersion: number;
  runId: string;
  videoId: string;
  navEpoch: number;
  errorCode: ErrorCode;
  message: string;
}

export interface PingMessage {
  type: "ping";
  protocolVersion: number;
  runId: string;
}

export type ExtensionMessage =
  | HelloMessage
  | PayloadMessage
  | VideoDoneMessage
  | VideoFailedMessage
  | PingMessage;

// -- Controller -> extension ---------------------------------------------

export interface CollectionConfig {
  maxRecommendations: number;
  scrollDelayMs: number;
  delayJitter: number;
  pageReadyTimeoutMs: number;
  ackTimeoutMs: number;
  maxScrollRounds: number;
  maxPageDurationMs: number;
  // When true, neither chain-exhaustion nor the recommendation cap ends
  // collection until the comments header's count is resolved (known to be
  // 0, or at least one thread has actually been captured) — otherwise a
  // low --max-recommendations (or a short recommendation chain) can end
  // the page before comments ever get a chance to load, since nothing
  // about the collector's own stopping conditions is comment-aware.
  // maxScrollRounds/maxPageDurationMs still bound this either way.
  waitForComments: boolean;
}

export interface HelloAckMessage {
  type: "hello_ack";
  protocolVersion: number;
  runId: string;
  videoId: string;
  videoUrl: string;
  config: CollectionConfig;
}

export interface PayloadAckMessage {
  type: "payload_ack";
  protocolVersion: number;
  batchId: string;
}

export interface PayloadNackMessage {
  type: "payload_nack";
  protocolVersion: number;
  batchId: string;
  retryable: boolean;
}

export interface VideoDoneAckMessage {
  type: "video_done_ack";
  protocolVersion: number;
  videoId: string;
}

export interface VideoFailedAckMessage {
  type: "video_failed_ack";
  protocolVersion: number;
  videoId: string;
}

export interface PongMessage {
  type: "pong";
  protocolVersion: number;
}

export interface StopMessage {
  type: "stop";
  protocolVersion: number;
}

/**
 * Synthesized locally by the native-host bridge when the controller socket
 * is unreachable (src/scraper/native_host.py, CONTROLLER_UNAVAILABLE_MESSAGE_TYPE).
 * Never sent by the controller itself — it can't be, by definition, when
 * this fires. Documented here because the service worker must recognise it
 * on the same native port as every other controller-originated message.
 */
export interface ControllerUnavailableMessage {
  type: "controller_unavailable";
  protocolVersion: number;
}

export type ControllerMessage =
  | HelloAckMessage
  | PayloadAckMessage
  | PayloadNackMessage
  | VideoDoneAckMessage
  | VideoFailedAckMessage
  | PongMessage
  | StopMessage
  | ControllerUnavailableMessage;

export function buildHello(restart: boolean): HelloMessage {
  return { type: "hello", protocolVersion: PROTOCOL_VERSION, restart };
}

export function buildPayload(fields: Omit<PayloadMessage, "type" | "protocolVersion">): PayloadMessage {
  return { type: "payload", protocolVersion: PROTOCOL_VERSION, ...fields };
}

export function buildVideoDone(
  fields: Omit<VideoDoneMessage, "type" | "protocolVersion">,
): VideoDoneMessage {
  return { type: "video_done", protocolVersion: PROTOCOL_VERSION, ...fields };
}

export function buildVideoFailed(
  fields: Omit<VideoFailedMessage, "type" | "protocolVersion">,
): VideoFailedMessage {
  return { type: "video_failed", protocolVersion: PROTOCOL_VERSION, ...fields };
}

export function buildPing(runId: string): PingMessage {
  return { type: "ping", protocolVersion: PROTOCOL_VERSION, runId };
}
